hoomd/box.py

```diff
@@ -0,0 +1,393 @@
+      Ly=aspect, dimensions=2, volume=A)`
+    * Cubic box with given length: `data.boxdim(L=10)`
+    * Fully define all box parameters: `data.boxdim(Lx=10, Ly=20, Lz=30, xy=1.0,
+      xz=0.5, yz=0.1)`
+    """
+
+    def __init__(self, Lx=1.0, Ly=1.0, Lz=1.0, xy=0.0, xz=0.0, yz=0.0):
+        self._cpp_obj = _hoomd.BoxDim(Lx, Ly, Lz)
+        self._cpp_obj.setTiltFactors(xy, xz, yz)
+        self._lattice_vectors = LatticeVectors(self._cpp_obj)
+
+    # Dimension based properties
+    @property
+    def dimension(self):
+        return 2 if self.Lz == 0 else 3
+
+    @property
+    def is2D(self):
+        return self.Lz == 0
+
+    # Length based properties
+    @property
+    def L(self):
+        return to_three_array(self._cpp_obj.getL())
+
+    @L.setter
+    def L(self, new_L):
+        if len(new_L) == 2:
+            new_L = [new_L[0], new_L[1], 0]
+        self._cpp_obj.setL(make_scalar3(new_L))
+
+    @property
+    def Lx(self):
+        return self.L[0]
+
+    @Lx.setter
+    def Lx(self, value):
+        L = self._cpp_obj.getL()
+        L.x = float(value)
+        self.L = L
+
+    @property
+    def Ly(self):
+        return self.L[1]
+
+    @Ly.setter
+    def Ly(self, value):
+        L = self._cpp_obj.getL()
+        L.y = float(value)
+        self.L = L
+
+    @property
+    def Lz(self):
+        return self.L[2]
+
+    @Lz.setter
+    def Lz(self, value):
+        L = self._cpp_obj.getL()
+        L.z = float(value)
+        self.L = L
+
+    # Box tilt based properties
+    @property
+    def tilts(self):
+        return np.array([self.xy, self.xz, self.yz])
+
+    @tilts.setter
+    def tilts(self, new_tilts):
+        if len(new_tilts) == 2:
+            if self.is2D:
+                self._cpp_obj.setTiltFactors(new_tilts[0], new_tilts[1], 0)
+            else:
+                raise ValueError("Must specify 3 tilt factors for 3D box.")
+        elif len(new_tilts) == 3:
+            self._cpp_obj.setTiltFactors(new_tilts[0], new_tilts[1],
+                                         new_tilts[2])
+        else:
+            raise ValueError("Tilt array must be either the box dimension "
+                             "or three.")
+
+    @property
+    def xy(self):
+        return self._cpp_obj.getTiltFactorXY()
+
+    @xy.setter
+    def xy(self, xy):
+        self.tilts = [xy, self.xz, self.yz]
+
+    @property
+    def xz(self):
+        return self._cpp_obj.getTiltFactorXZ()
+
+    @xz.setter
+    def xz(self, xz):
+        self.tilts = [self.xy, xz, self.yz]
+
+    @property
+    def yz(self):
+        return self._cpp_obj.getTiltFactorYZ()
+
+    @yz.setter
+    def yz(self, yz):
+        self.tilts = [self.xy, self.xz, yz]
+
+    # Misc. properties
+    @property
+    def periodicity(self):
+        return to_three_array(self._cpp_obj.getPeriodic())
+
+    @periodicity.setter
+    def periodicity(self, value):
+        '''Set the periodic versus closed boundary conditions'''
+        # TODO: check that this works
```

Comment:
The C++ code does assert that all 3 directions are periodic. It may be possible to relax this requirement. I don't know how much work that would take.

Reply:
Ah, well. I can at least leave the getter for completeness.